File: C Sharp/Object Validator/Runtime/Contains.cs
using System.Collections;

namespace Quicksilver.ObjectValidator.Runtime;
class Contains(bool reversed, string? fieldExpression, string? arg) : Condition(reversed, fieldExpression)
{
    private readonly string? arg = arg;

    protected override bool IsFulfilledBy(object? value, string? fullFieldExpression, HashSet<string> passedFields, HashSet<string> failedFields)
    {
        switch (value) {
            case null:
                throw new Exception("Null values are not supported for 'contains'.");
            case string s:
                return s.Contains(arg!);
            case IEnumerable e:
                foreach (object o in e)
                    if (o == null) {
                        if (arg == null)
                            return true;
                    } else if (o.ToString() == arg)
                        return true;
                return false;
            default:
                throw new Exception("Unsupported type for 'contains': " + value.GetType());
        }
    }
}

File: Java/Object Validator/lib/src/main/java/com/quicksilver/objectvalidator/runtime/Contains.java
package com.quicksilver.objectvalidator.runtime;

import java.util.HashSet;

public class Contains extends Condition {
    private final String arg;

    public Contains(boolean reversed, String fieldExpression, String arg) {
        super(reversed, fieldExpression);
        this.arg = arg;
    }

    @Override
    protected boolean isFulfilledBy(Object value, String fullFieldExpression, HashSet<String> passedFields, HashSet<String> failedFields) {
        return switch (value) {
            case null -> throw new RuntimeException("Null values are not supported for 'contains'.");
            case String s -> s.contains(arg);
            case Object[] a -> {
                for (Object o : a)
                    if (o == null) {
                        if (arg == null)
                            yield true;
                    } else if (o.toString().equals(arg))
                        yield true;
                yield false;
            }
            case Iterable<?> i -> {
                for (Object o : i)
                    if (o == null) {
                        if (arg == null)
                            yield true;
                    } else if (o.toString().equals(arg))
                        yield true;
                yield false;
            }
            default -> throw new RuntimeException("Unsupported type for 'contains': " + value.getClass());
        };
    }
}
